### ISSUE TITLE
PKCS#11 support with HAProxy 

### ISSUE THREAD

**Issue by type4ranjan** [haproxy/haproxy#71]
```
(paste your output here)
```

## What should haproxy do differently? Which functionality do you think we should add?

Haproxy should have official support of PKCS#11 so that keys can be stored on HSM and used for TLS.

## What are you trying to do?

Haproxy should have official support of PKCS#11 so that keys can be stored on HSM and used for TLS offloading.I have tried integrating it with Safenet HSM but still waiting for the official release.

**Comment by git001** [comment]
Is the `openssl engine` https://cbonte.github.io/haproxy-dconv/1.9/configuration.html#ssl-engine with https://github.com/OpenSC/libp11 not a possible solution for your use case?

Maybe this issue thread help you to create a solution https://github.com/OpenSC/libp11/issues/160

**Comment by type4ranjan** [comment]
Yes providing ssl-engine in haproxy.cfg file is a solution but i haven't find any official documentation/guidelines  from HAProxy for stepwise instructions and configuration of HAProxy to work with HSM.Request you to release it officially.

**Comment by type4ranjan** [comment]
Dear @wlallemand,
Any update on it when it will be supported. Thanks

**Comment by wlallemand** [comment]
No plan was done for the pkcs11 yet. It wasn't ready when we worked on the OpenSSL 3 Engine support.

**Comment by wlallemand** [comment]
No plan was done for the pkcs11 yet. The pkcs11 provider wasn't ready when we worked on the OpenSSL 3 providers support in HAProxy.

However the TPM2 module which was mentioned in #518 was tested sucessfully. A documentation is available on the wiki: 

https://github.com/haproxy/wiki/wiki/OpenSSL-Providers-in-HAProxy

If you are aware of another HSM module compatible with the OpenSSL 3 providers API there is a chance it could work without any changes in HAProxy.

**Comment by type4ranjan** [comment]
The HSM module we are using has support with OpenSSL 1.1.1 and it uses OpenSC Libp11. 
OpenSC is not yet fully supported with OpenSSL 3. 
Can you please plan for its support it so that it can load key from any hsm module through pkcs11 engine?

**Comment by wlallemand** [comment]
It's not planned for now. If you wish to work on this, we will be happy to integrate contributions on the subject. HAProxy >= 2.6 still supports the engine API, but it needs to be compiled with USE_ENGINE=1. But the Engine API is officially deprecated in the latest OpenSSL release.

**Comment by type4ranjan** [comment]
I can help in testing this feature with HSM module if someone can integrate.

**Comment by git001** [comment]
@type4ranjan maybe you can create a patch and contribute to HAProxy :smile:   
Please take a look into the https://github.com/haproxy/haproxy/blob/master/CONTRIBUTING file for contribution

**Comment by tmccombs** [comment]
> Is the openssl engine https://cbonte.github.io/haproxy-dconv/1.9/configuration.html#ssl-engine with https://github.com/OpenSC/libp11 not a possible solution for your use case?

Is there an example of how to use that?

I figure I would use `ssl-engine pkcs11` or maybe `ssl-engine pkcs#11`, but what do I use for the `crt` file?

**Comment by simo5** [comment]
@tmccombs I would look at the pkcs11 provider, given engines are now deprecated in OpenSSL 1.1 and that version of OpenSSL will rapidly be EOL as well... would it be hard for HAProxy to use the OpenSSL 3 STORE API and allow specifying a PKCS#11 URI instead of a file name for certificates/keys ?

**Comment by quality-leftovers** [comment]
For anyone reading this: Implementation of key engine / OpenSSL store is mostly blocked by current configuration format, which does not allow specifying keys for a certificate separately but infers the name of the key file name based on the name of the certificate file name (and some settings) => see #785.

**Comment by simo5** [comment]
At least for pkcs11 URIs this is not a huge problem, as customarily certificate objects and key objects can share the same CKA_ID (or sometime the same label). And OpenSSL knows what kind of object it wants so it is generally ok to avoid specifying the type attribute in the URI.
Ideally this situation should be improved though, forcing filenames is not a great experience in many situations.

**Comment by simo5** [comment]
Ideally you can mix and match filenames/URIs for keys and certs
sometimes you may have tokens that can store keys but not certs, or other things like that.

Cert and key storage should be unrelated.

**Comment by wlallemand** [comment]
I agree that ideally the cert and key must not be in the same file. This was a wrong decision that was taken at the beginning of the SSL implementation, so we are kind of stuck with the existing configuration. Unfortunately the key and cert separation has not yet landed in haproxy, because we have huge problems of backward compatibilities and it breaks a lot of the CLI... So we are still working on it.

Regarding the pkcs11, there is a mecanism in haproxy which builds a cache for the certificates, so they can be modified dynamically, the loader needs to be able to work with buffers and files.  I don't know if we can easilly integrate the OSSL_STORE_open() function within the parser, I don't think so. However, maybe we could make an exception in the parser if we find a pkcs11 URI and uses the OSTORE API in this case.

**Comment by tmccombs** [comment]
> I would look at the pkcs11 provider

Is there any way to use that from haproxy right now?

**Comment by simo5** [comment]
@tmccombs I assume the problem is passing the pkcs11 URI through the haproxy config, if tat can be resolved the changes needed to make use of the pkcs11-provider should be small.

**Comment by ojw-nc** [comment]
Is this now possible since recent work on #785, or something still required outstanding?

**Comment by wlallemand** [comment]
It still lacks a pkcs11 uri config param. But since https://github.com/latchset/pkcs11-provider/issues/222 the pkcs11 provider supports the pkcs11 URI inside a PEM, it should work with HAProxy since v0.4 ( https://github.com/latchset/pkcs11-provider/releases/tag/v0.4 )

There is already a documentation on how to do this here: https://github.com/haproxy/wiki/wiki/OpenSSL-Providers-in-HAProxy#pkcs11-provider

The discussion started from the mailing-list:  https://www.mail-archive.com/haproxy@formilux.org/msg44702.html
The only caveat seems to be some incompatibility with the master-worker mode, which need to be investigated.

**Comment by chrisstaite-menlo** [comment]
I've been considering this.  I wonder, is there appetite for adding a build flag to HAProxy which would enable code that ties together p11-kit (https://p11-glue.github.io/p11-glue/p11-kit.html) and `ssl_sock_put_srv_ckch_into_ctx` in HAProxy such that if a private key file is a `PKCS#11 PROVIDER URI` it looks it up in p11-kit?

My proposal is that enabling this build flag (USE_P11_KIT) it would link HAProxy against p11-kit-client.so, enable a new global configuration variable (p11-kit-address) which sets the address of the p11-kit server.  If this address is set and a PEM is loaded with `PKCS#11 PROVIDER URI`, then a shim will be created around the p11-kit-client to create an `SSL_PRIVATE_KEY_METHOD` structure which is set on the context using `SSL_CTX_set_private_key_method`.  This should work with AWS-LC, BoringSSL and OpenSSL 1.1.

For OpenSSL 3.0 there is already support in the pkcs11-provider as noted in the documentation linked above.  I assume there would be a way to get this to work for WolfTLS also, but I'm not sure what that is yet.

Happy to put together a patch set for this if it would be appropriate to merge to HAProxy in this state.

**Comment by kubrickfr** [comment]
For me the main interest is that if you do that, you open up use cases for things like having your certificates stored in EC2 secure enclaves, and therefore the ability to use ACM certificates.

**Comment by chrisstaite-menlo** [comment]
Yes, the purpose for it would be to provide access to CloudHSM (AWS) and CloudKMS (GCP).

**Comment by simo5** [comment]
Would this conflict with using pkcs11_uris with openssl ?

**Comment by chrisstaite-menlo** [comment]
Firstly, you'd have to compile it in, and secondly it would only be enabled if you configured `p11-kit-address`, so I don't see that it could conflict with anything.

**Comment by wlallemand** [comment]
I'm not sure p11-glue would help with anything, the problem is that the "crt" parameter in haproxy is a certificate identifier, which is used as a key in a tree, and used elsewhere in the code, because there is an internal cache for certificates. 

Previously this filename was abused to pass pkcs11 URIs but this was not expected to be used like this. These parameters are still usable in PEM files with HSMs we are seeing in the field, so this is not really a requirement for HSMs.

However since we worked on splitting the parameters of certificates since the `crt-store` keyword, crt and key are separated parameters, we could have new parameters in the future, but nothing's planned yet regarding this.

**Comment by chrisstaite-menlo** [comment]
Yup, I realised that when I went to look in to it.  I've done some prototyping and have come up with something.  The major downside is that the PKCS#11 interface is synchronous.  I'll report back when I've completed my prototyping to see what the thoughts are.  Essentially, I've now got a thread pool for signing.

**Comment by chrisstaite-menlo** [comment]
I've submitted a WIP patch up to the mailing list in case anyone is interested in taking it for a spin.
